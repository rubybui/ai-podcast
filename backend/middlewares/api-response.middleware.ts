import { Request, Response, NextFunction } from "express";
import { StatusCodeEnum } from "../enums";

export interface IPagination {
  currentPage: number;
  perPage: number;
  total: number;
  totalPages: number;
};
export interface IApiResponse {
  success: boolean;
  message?: string;
  data?: any;
  pagination? 
}

export interface ICustomResponse extends Response {
  success: (params: {
    message?: string;
    data?: any;
    pagination?: IPagination;
    statusCode?: number;
  }) => any;
  error: (params: { message?: string; statusCode?: number }) => any;
}

export function customApiResponse(
  req: Request,
  res: ICustomResponse,
  next: NextFunction
) {
  res.success = ({ message, data, pagination, statusCode }) => {
    const response: IApiResponse = {
      success: true,
    };
    if (message) response.message = message;
    if (data) response.data = data;
    if (pagination) response.pagination = pagination;
    res.status(statusCode || StatusCodeEnum.OK).json(response);
  };

  res.error = ({ message, statusCode }) => {
    const response: IApiResponse = {
      success: false,
    };
    if (message) response.message = message;
    res.status(statusCode || StatusCodeEnum.BAD_REQUEST).json(response);
  };

  next();
}
