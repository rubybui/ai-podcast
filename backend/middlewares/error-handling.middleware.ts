import { ErrorRequestHandler } from "express";
import { StatusCodeEnum } from "../enums";
import { ICustomResponse } from "./api-response.middleware";

export const controllerWrapper = (fn) => (req, res, next) =>
  Promise.resolve(fn(req, res, next)).catch((err) => next(err));

export const errorHandling: ErrorRequestHandler = (
  err: any,
  req,
  res: ICustomResponse,
  next
) => {
  console.log(err);
  return res.error({
    message: "Something went wrong",
    statusCode: StatusCodeEnum.INTERNAL_ERROR,
  });
};
