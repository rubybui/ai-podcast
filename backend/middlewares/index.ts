export * from "./api-response.middleware";
export * from "./validate-request.middleware";
export * from "./upload-file.middleware";
export * from "./error-handling.middleware";
