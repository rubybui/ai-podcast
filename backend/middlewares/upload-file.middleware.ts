import * as multer from "multer";

export const uploadMiddleware = multer({
  limits: {
    fileSize: 4 * 1024 * 1024,
  },
});
