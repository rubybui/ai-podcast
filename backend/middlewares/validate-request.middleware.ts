import { Request, Response, NextFunction } from "express";
import { validateSync } from "class-validator";
import { StatusCodeEnum } from "../enums";

export function validateRequest(dto: any) {
  return (req: Request, res: Response, next: NextFunction) => {
    const errors = validateSync(
      Object.assign(new dto(), { ...req.body, ...req.query })
    );

    if (errors.length > 0) {
      const constraintError = errors[0].constraints;
      return res.status(StatusCodeEnum.BAD_REQUEST).json({
        statusCode: StatusCodeEnum.BAD_REQUEST,
        message: constraintError[Object.keys(constraintError)[0]],
      });
    }

    next();
  };
}
