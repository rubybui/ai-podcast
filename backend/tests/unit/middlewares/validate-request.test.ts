import { IsNotEmpty } from "class-validator";
import { validateRequest } from "../../../middlewares";
import { StatusCodeEnum } from "../../../enums";

class MockedDto {
  @IsNotEmpty()
  firstName: string;

  @IsNotEmpty()
  lastName: string;
}

const mockReq: any = {
  body: {
    firstName: "Rio",
    lastName: "Test",
  },
};

describe("validateRequest() middleware", () => {
  let mockNext: any, mockRes: any;
  beforeEach(() => {
    jest.resetAllMocks();
    mockNext = jest.fn();
    mockRes = {
      status: jest.fn().mockImplementation((number) => ({
        json: (data) => data,
      })),
    };
  });

  describe("success", () => {
    test("should call next if request request body is valid", () => {
      validateRequest(MockedDto)(mockReq, mockRes, mockNext);
      expect(mockRes.status).toBeCalledTimes(0);
      expect(mockNext).toBeCalledTimes(1);
    });

    test("should call next if request query is valid", () => {
      const mockReqQuery: any = { query: mockReq.body };
      validateRequest(MockedDto)(mockReqQuery, mockRes, mockNext);
      expect(mockRes.status).toBeCalledTimes(0);
      expect(mockNext).toBeCalledTimes(1);
    });
  });

  describe("error", () => {
    test("should throw error", () => {
      const body: any = {};
      const response: any = validateRequest(MockedDto)(body, mockRes, mockNext);
      expect(response.statusCode).toBe(StatusCodeEnum.BAD_REQUEST);
      expect(response).toMatchSnapshot();
      expect(mockRes.status).toBeCalledTimes(1);
      expect(mockNext).toBeCalledTimes(0);
    });
  });
});
