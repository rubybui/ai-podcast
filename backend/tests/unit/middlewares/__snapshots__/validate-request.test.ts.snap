// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateRequest() middleware error should throw error 1`] = `
{
  "message": "firstName should not be empty",
  "statusCode": 400,
}
`;
